turbo/snapshotsync/snapshotsync.go

```diff
@@ -66,11 +70,182 @@ func RequestSnapshotsDownload(ctx context.Context, downloadRequest []services.Do
+	}
+	stepPrune = adjustStepPrune(stepPrune)
+	blockPrune = adjustBlockPrune(blockPrune, minBlockToDownload)
+	snapshotKindToNames := make(map[string][]snapshotFileData)
+	for _, p := range preverified {
+		name := p.Name
+		// Dont prune unprunable files
+		if !canSnapshotBePruned(name) {
+			continue
+		}
+		var from, to uint64
+		var err error
+		var kind string
+		if shouldUseStepsForPruning(name) {
+			// parse "from" (0) and "to" (64) from the name
+			// parse the snapshot "kind". e.g kind of 'idx/v1-accounts.0-64.ef' is "idx/v1-accounts"
+			rangeString := strings.Split(name, ".")[1]
+			rangeNums := strings.Split(rangeString, "-")
+			// convert the range to uint64
+			from, err = strconv.ParseUint(rangeNums[0], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			to, err = strconv.ParseUint(rangeNums[1], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			kind = strings.Split(name, ".")[0]
+		} else {
+			// e.g 'v1-000000-000100-beaconblocks.seg'
+			// parse "from" (000000) and "to" (000100) from the name. 100 is 100'000 blocks
+			minusSplit := strings.Split(name, "-")
+			s, _, _ := snaptype.ParseFileName("tmp", name)
+			from = s.From
+			to = s.To
+			kind = minusSplit[3]
+		}
+		blackList[p.Name] = struct{}{} // Add all of them to the blacklist and remove the ones that are not blacklisted later.
+		snapshotKindToNames[kind] = append(snapshotKindToNames[kind], snapshotFileData{
+			from:      from,
+			to:        to,
+			stepBased: shouldUseStepsForPruning(name),
+			name:      name,
+		})
+	}
+	// sort the snapshots by "from" and "to" in ascending order
+	for _, snapshots := range snapshotKindToNames {
+		prunedDistance := uint64(0) // keep track of pruned distance for snapshots
+		// sort the snapshots by "from" and "to" in descending order
+		sort.Slice(snapshots, func(i, j int) bool {
+			if snapshots[i].from == snapshots[j].from {
+				return snapshots[i].to > snapshots[j].to
+			}
+			return snapshots[i].from > snapshots[j].from
+		})
+		for _, snapshot := range snapshots {
+			if snapshot.stepBased {
+				if prunedDistance >= uint64(stepPrune) {
+					break
+				}
+			} else if prunedDistance >= uint64(blockPrune) {
+				break
+			}
+
+			if snapshot.stepBased {
+				delete(blackList, snapshot.name)
+				prunedDistance += snapshot.to - snapshot.from
+				continue
+			}
+			delete(blackList, snapshot.name)
+			prunedDistance += snapshot.to - snapshot.from
+		}
+	}
+	return blackList, nil
+}
+
+// getMinimumBlocksToDownload - get the minimum number of blocks to download
+func getMinimumBlocksToDownload(tx kv.Tx, blockReader services.FullBlockReader, minStep uint64, expectedPruneBlockAmount uint64) (uint64, uint64, error) {
+	frozenBlocks := blockReader.Snapshots().SegmentsMax()
+	minToDownload := uint64(math.MaxUint64)
+	minStepToDownload := minStep
+	stateTxNum := minStep * config3.HistoryV3AggregationStep
+	if err := blockReader.IterateFrozenBodies(func(blockNum, baseTxNum, txAmount uint64) error {
```

Comment:
seems you only need download bodies.seg - without headers.seg?

Reply:
Yes, maybe I can only download `bodies.seg`. but I would have to download full header chain anyway later, it also make it easy to name it as header chain sync